xpra/platform/posix/keyboard.py

```diff
@@ -93,7 +92,7 @@ def set_platform_layout(self, layout: str) -> None:
         if index is None:
             log(f"asked layout ({layout}) has no corresponding registered input source")
             return
-        self._dbus_gnome_shell_eval_ism(f".inputSources[{index}].activate()")
+        self._dbus_gnome_shell_ism("Activate", [native_to_dbus(index)])
```

Comment:
A better approach to these chunks would be to test which interface is available and then choose the most appropriate one.

Reply:
Good idea. Updated. Now, it falls back to the **Eval** method and works as while as one of these conditions is true:

1. new extension is enabled by end-user,
2. the GNOME Shell is running in **unsafe** mode,
3. the GNOME Shell is older than v41.0.

Reply:
I must be missing something because I can't see where it does the fallback.

Reply:
Code runs asynchronously, sending a D-Bus request and calling one of the `ok_cb` or `err_cb` functions based on the successful or failed invocation. So, if extension is not installed or is not enabled, `err_cb` will be called which will send the fallback D-Bus request. It's around lines 75-85:
```python
        def err_cb(msg) -> None:
            log("_dbus_gnome_shell_ism(method=%s, args=%s): %s", method, args, msg)
            if method == "List":
                self._dbus_gnome_shell_eval_ism(".inputSources", self._store_input_sources)
            elif method == "Activate":
                try:
                    index = dbus_to_native(args[0])
                except Exception:
                    log("_dbus_gnome_shell_ism(method=%s, args=%s): the Eval fallback", method, args, exc_info=True)
                else:
                    self._dbus_gnome_shell_eval_ism(f".inputSources[{index}].activate()")
```

Reply:
Ah, thanks, now I see it.